gapis/api/vulkan/transform_command_disabler.go

```diff
@@ -0,0 +1,418 @@
+		append(api.SubCmdIdx{}, id...))
+	return nil
+}
+
+func (disablerTransform *commandDisabler) RequiresAccurateState() bool {
+	return false
+}
+
+func (disablerTransform *commandDisabler) RequiresInnerStateMutation() bool {
+	return true
+}
+
+func (disablerTransform *commandDisabler) SetInnerStateMutationFunction(mutator transform.StateMutator) {
+	disablerTransform.stateMutator = mutator
+}
+
+func (disablerTransform *commandDisabler) BeginTransform(ctx context.Context, inputState *api.GlobalState) error {
+	disablerTransform.allocations = NewAllocationTracker(inputState)
+	return nil
+}
+
+func (disablerTransform *commandDisabler) EndTransform(ctx context.Context, inputState *api.GlobalState) ([]api.Cmd, error) {
+	return nil, nil
+}
+
+func (disablerTransform *commandDisabler) ClearTransformResources(ctx context.Context) {
+	disablerTransform.allocations.FreeAllocations()
+}
+
+func (disablerTransform *commandDisabler) TransformCommand(ctx context.Context, id transform.CommandID, inputCommands []api.Cmd, inputState *api.GlobalState) ([]api.Cmd, error) {
+	if len(inputCommands) == 0 {
+		return inputCommands, nil
+	}
+
+	if id.GetCommandType() != transform.TransformCommand {
+		// We do not interest with the artificial commands from endTransform.
+		return inputCommands, nil
+	}
+
+	currentSubCmdID := api.SubCmdIdx{uint64(id.GetID())}
+	if !disablerTransform.shouldBeDisabled(currentSubCmdID) {
+		return inputCommands, nil
+	}
+
+	queueSubmitProcessed := false
+	for _, cmd := range inputCommands {
+		if queueSubmitCmd, ok := cmd.(*VkQueueSubmit); ok {
+			if queueSubmitProcessed {
+				panic("We should not have more than one vkQueueSubmit for a single command")
+			}
+
+			queueSubmitProcessed = true
+			if err := disablerTransform.removeCommandFromVkQueueSubmit(ctx, currentSubCmdID, queueSubmitCmd, inputState); err != nil {
+				log.E(ctx, "Failed during removing command from VkQueueSubmit : %v", err)
+				return nil, err
+			}
+		} else {
+			if err := disablerTransform.writeCommand(cmd); err != nil {
+				log.E(ctx, "Failed during processing input commands : %v", err)
+				return nil, err
+			}
+		}
+	}
+
+	if !queueSubmitProcessed {
+		return nil, fmt.Errorf("No queue submit has found in command path")
+	}
+
+	return nil, nil
+}
+
+func (disablerTransform *commandDisabler) removeCommandFromVkQueueSubmit(ctx context.Context, idx api.SubCmdIdx, cmd *VkQueueSubmit, inputState *api.GlobalState) error {
+	layout := inputState.MemoryLayout
+	cb := CommandBuilder{Thread: cmd.Thread(), Arena: inputState.Arena}
+	cmd.Extras().Observations().ApplyReads(inputState.Memory.ApplicationPool())
+
+	submitInfos := cmd.PSubmits().Slice(0, uint64(cmd.SubmitCount()), layout).MustRead(ctx, cmd, inputState, nil)
+	newSubmitInfos := []VkSubmitInfo{}
+
+	newSubmit := cb.VkQueueSubmit(cmd.Queue(), cmd.SubmitCount(), cmd.PSubmits(), cmd.Fence(), cmd.Result())
+	newSubmit.Extras().MustClone(cmd.Extras().All()...)
+
+	for i := 0; i < len(submitInfos); i++ {
```

Comment:
Correct me if I wrong but afaik when you use range in that way, it copies all the object in range. It does not matter for the Vulkan handles but in terms of submit info, it actually may introduce copying the all struct. Maybe we can get some input from @pmuetschard. Because there are not many documentation about it and I am not very comfortable about my knowledge on the topic.

Reply:
Oh ok you want to avoid the copies. Is it really significant in terms of performance? Since it is not the canonical way in Golang, I would add a comment to say something like "// use range only for indexes to avoid copy", such that when someone else sees this code they understand why it is this way.

Reply:
Actually, here you have a copy anyway as `submitInfos[i]` is used as a function call argument, e.g. in `append(newSubmitInfos, submitInfos[i])` and `removeCommandFromSubmit(..., submitInfos[i], ...)`.
So you might as well use `i, submitInfo := range submitInfos`?

Reply:
Might be.